### PR TITLE
Use safevalues to fix trusted types issues reported by tsec

### PR DESCRIPTION
**Pull request by dlarocque**
We recently added [tsec](https://github.com/google/tsec) in #8285, to report errors where code is vulnerable to XSS. In this PR, we begin resolving some of the reported errors by introducing [safevalues](https://github.com/google/safevalues) and apply it to fix a few of the simpler cases, where URLs must be wrapped in a `trustedResourceUrl`. For cases that are more complex to solve, I have added `FIXME`'s noting that they must be fixed using the safevalues library.

**Testing**: I have added assertions to our existing tests to ensure that the scripts' URLs are correctly assigned. For an additional sanity check, I've also created a React app using Firebase with safevalues, and verified that the `trustedResourceUrl`'s in App Check are actually attaching the `script` to the DOM with the correct URL using the Browser DevTools and just stepping through the code and observing state.

### CODE REVIEW

File: packages/messaging/src/helpers/registerDefaultSw.ts [hunk 1]
```diff
@@ -24,8 +27,10 @@ export async function registerDefaultSw(
   messaging: MessagingService
 ): Promise<void> {
   try {
-    messaging.swRegistration = await navigator.serviceWorker.register(
-      DEFAULT_SW_PATH,
+    const container = navigator.serviceWorker;
+    messaging.swRegistration = await safeServiceWorkerContainer.register(
+      container,
+      trustedResourceUrl`/firebase-messaging-sw.js`,
```
Comment (dlarocque) [inline packages/messaging/src/helpers/registerDefaultSw.ts:33]: This is used in `getToken` and `deleteToken`.

The tests are passing. Is that enough to be confident that this works?

File: packages/database/src/realtime/BrowserPollConnection.ts [hunk 2]
```diff
@@ -475,7 +477,7 @@ export class FirebaseIFrameScriptHolder {
       const iframeContents = '<html><body>' + script + '</body></html>';
       try {
         this.myIFrame.doc.open();
-        this.myIFrame.doc.write(iframeContents);
+        safeDocument.write(this.myIFrame.doc, sanitizeHtml(iframeContents));
```
Comment (dlarocque) [inline packages/database/src/realtime/BrowserPollConnection.ts:480]: Again, I would like to verify that this actually works, but I am not sure how to do that beyond running the tests.

**Review by dlarocque — COMMENTED** [review]
I am confident that our usage of `trustedResourceUrl` works, but I feel that I need to spend much more time looking into how I can test `write` and `safeServiceWorkerContainer.register`, as those seem more complex and issues may not be caught by the tests. Any suggestions on how I can test these two would be very helpful.

**Review by hsubox76 — APPROVED** [review]
Looks good! I think we can make the FIXMEs into TODOs for easy context for whoever works on the fix, and set up time to work with the product teams on each of the more complicated cases. We can bring this up as an FYI in the biweekly meeting and contributors chat.

**Comment by nicole0707** [comment]
Hi team, we are upgrading Firebase sdk to `10.12.4`   and encountered some issues in our unit tests. Could you please help us to resolve them? Thanks

```tsx
    TypeError: 
        ############################## ERROR ##############################

        It looks like you are trying to call a template tag function (fn`...`)
        using the normal function syntax (fn(...)), which is not supported.

        The functions in the safevalues library are not designed to be called
        like normal functions, and doing so invalidates the security guarantees
        that safevalues provides.

        If you are stuck and not sure how to proceed, please reach out to us
        instead through:
         - https://github.com/google/safevalues/issues

        ############################## ERROR ##############################

      30 |     const appCheckProvider = new ReCaptchaEnterpriseProvider(siteKey);
      31 |
    > 32 |     const appCheck = initializeAppCheck(firebaseApp, {
         |                                        ^
      33 |       provider: appCheckProvider,
      34 |       isTokenAutoRefreshEnabled: true,
      35 |     });

      at assertIsTemplateObject (../../node_modules/.pnpm/safevalues@0.6.0/node_modules/safevalues/dist/cjs/internals/string_literal.js:19:15)
      at Object.trustedResourceUrl (../../node_modules/.pnpm/safevalues@0.6.0/node_modules/safevalues/dist/cjs/builders/resource_url_builders.js:162:56)
      at trustedResourceUrl (../../node_modules/.pnpm/@firebase+app-check@0.8.6_@firebase+app@0.10.7/node_modules/@firebase/app-check/src/recaptcha.ts:186:5)
      at initializeEnterprise (../../node_modules/.pnpm/@firebase+app-check@0.8.6_@firebase+app@0.10.7/node_modules/@firebase/app-check/src/recaptcha.ts:72:5)
      at ReCaptchaEnterpriseProvider.initializeRecaptchaEnterprise [as initialize] (../../node_modules/.pnpm/@firebase+app-check@0.8.6_@firebase+app@0.10.7/node_modules/@firebase/app-check/src/providers.ts:209:5)
      at _activate (../../node_modules/.pnpm/@firebase+app-check@0.8.6_@firebase+app@0.10.7/node_modules/@firebase/app-check/src/api.ts:160:18)
      at initializeAppCheck (../../node_modules/.pnpm/@firebase+app-check@0.8.6_@firebase+app@0.10.7/node_modules/@firebase/app-check/src/api.ts:106:3)
      at src/app-check/app-check.tsx:32:40
      at apply (../utils/src/memo.tsx:41:29)
      at getAppCheckToken (src/app-check/app-check-token.tsx:23:53)
      at Object.<anonymous> (src/app-check/app-check-token.test.tsx:92:36)
```

**Comment by dlarocque** [comment]
> Hi team, we are upgrading Firebase sdk to `10.12.4` and encountered some issues in our unit tests. Could you please help us to resolve them? Thanks
> 
> ```tsx
>     TypeError: 
>         ############################## ERROR ##############################
> 
>         It looks like you are trying to call a template tag function (fn`...`)
>         using the normal function syntax (fn(...)), which is not supported.
> 
>         The functions in the safevalues library are not designed to be called
>         like normal functions, and doing so invalidates the security guarantees
>         that safevalues provides.
> 
>         If you are stuck and not sure how to proceed, please reach out to us
>         instead through:
>          - https://github.com/google/safevalues/issues
> 
>         ############################## ERROR ##############################
> 
>       30 |     const appCheckProvider = new ReCaptchaEnterpriseProvider(siteKey);
>       31 |
>     > 32 |     const appCheck = initializeAppCheck(firebaseApp, {
>          |                                        ^
>       33 |       provider: appCheckProvider,
>       34 |       isTokenAutoRefreshEnabled: true,
>       35 |     });
> 
>       at assertIsTemplateObject (../../node_modules/.pnpm/safevalues@0.6.0/node_modules/safevalues/dist/cjs/internals/string_literal.js:19:15)
>       at Object.trustedResourceUrl (../../node_modules/.pnpm/safevalues@0.6.0/node_modules/safevalues/dist/cjs/builders/resource_url_builders.js:162:56)
>       at trustedResourceUrl (../../node_modules/.pnpm/@firebase+app-check@0.8.6_@firebase+app@0.10.7/node_modules/@firebase/app-check/src/recaptcha.ts:186:5)
>       at initializeEnterprise (../../node_modules/.pnpm/@firebase+app-check@0.8.6_@firebase+app@0.10.7/node_modules/@firebase/app-check/src/recaptcha.ts:72:5)
>       at ReCaptchaEnterpriseProvider.initializeRecaptchaEnterprise [as initialize] (../../node_modules/.pnpm/@firebase+app-check@0.8.6_@firebase+app@0.10.7/node_modules/@firebase/app-check/src/providers.ts:209:5)
>       at _activate (../../node_modules/.pnpm/@firebase+app-check@0.8.6_@firebase+app@0.10.7/node_modules/@firebase/app-check/src/api.ts:160:18)
>       at initializeAppCheck (../../node_modules/.pnpm/@firebase+app-check@0.8.6_@firebase+app@0.10.7/node_modules/@firebase/app-check/src/api.ts:106:3)
>       at src/app-check/app-check.tsx:32:40
>       at apply (../utils/src/memo.tsx:41:29)
>       at getAppCheckToken (src/app-check/app-check-token.tsx:23:53)
>       at Object.<anonymous> (src/app-check/app-check-token.test.tsx:92:36)
> ```

Hi @nicole0707, sorry that you're suddenly running into issues with this.

I just tried to re-create this error message in my own Firebase web app, but was unsuccessful.
I'd really like to figure out why you're seeing this issue, and reproduce it myself. It would be helpful if you could share: 
- Code snippet of your unit test that raised the error
- Your environment (any web frameworks, testing frameworks, libraries etc...)
- Your testing configuration

**Comment by nicole0707** [comment]
Hi @dlarocque, FYI we are using the Firebase sdk with Nextjs, and the testing framework is Jest. I will look into the issue and try to provide a minimal demo to reproduce it. Thanks!

**Comment by delmaass** [comment]
Hello Firebase team, my NextJS project won't build because of an exception that seem related to this update:

```
./node_modules/firebase/analytics/dist/index.mjs + 51 modules
Cannot get final name for export 'trustedResourceUrl' of ./node_modules/safevalues/dist/mjs/index.js
```

I am using `firebase: ^10.12.4`

**Comment by delmaass** [comment]
> Hello Firebase team, my NextJS project won't build because of an exception that seem related to this update:
> 
> ```
> ./node_modules/firebase/analytics/dist/index.mjs + 51 modules
> Cannot get final name for export 'trustedResourceUrl' of ./node_modules/safevalues/dist/mjs/index.js
> ```
> 
> I am using `firebase: ^10.12.4`

Downgrading to `"firebase": "10.12.3"` solved the issue

**Comment by dlarocque** [comment]
> Hello Firebase team, my NextJS project won't build because of an exception that seem related to this update:
> 
> ```
> ./node_modules/firebase/analytics/dist/index.mjs + 51 modules
> Cannot get final name for export 'trustedResourceUrl' of ./node_modules/safevalues/dist/mjs/index.js
> ```
> 
> I am using `firebase: ^10.12.4`

Hi @delmaass, thanks for reporting this. As a sanity check, I upgraded my own Next.js app (It's just the template app with basic Firebase usage) to Firebase 10.12.4, but I was not able to reproduce your issue when building or deploying.
If you believe this is an issue caused by the SDK, could you please submit a new Issue in this repo including more details?

I am worried this might cause issues for a lot of Next.js users, but I haven't been able to reproduce any issues myself.

**Comment by dlarocque** [comment]
@nicole0707 
I have been able to reproduce your issue with Jest. If you have any additional info that may help us solve this issue, please mention it in
https://github.com/firebase/firebase-js-sdk/issues/8386

**Comment by dlarocque** [comment]
> Hi @dlarocque, FYI we are using the Firebase sdk with Next.js, and the testing framework is Jest. I will look into the issue and try to provide a minimal demo to reproduce it. Thanks!

Hi @nicole0707, this should be fixed in the next release which will be out by the end of the week.

For more information, see #8395

**Comment by nicole0707** [comment]
> > Hi @dlarocque, FYI we are using the Firebase sdk with Next.js, and the testing framework is Jest. I will look into the issue and try to provide a minimal demo to reproduce it. Thanks!
> 
> Hi @nicole0707, this should be fixed in the next release which will be out by the end of the week.
> 
> For more information, see #8395

Thanks @dlarocque, I just checked the issue has been gone in 10.12.5.

**Pull request merged**